src/common/k8s-api/api-manager/api-manager.ts

```diff
@@ -21,87 +22,71 @@ export type KubeObjectStoreFrom<Api> = Api extends KubeApi<infer KubeObj, infer
   ? KubeObjectStore<KubeObj, Api, ApiData>
   : never;
 
+export type FindApiCallback = (api: KubeApi<KubeObject>) => boolean;
+
+interface Dependencies {
+  readonly apis: IComputedValue<KubeApi[]>;
+  readonly stores: IComputedValue<KubeObjectStore[]>;
+}
+
 export class ApiManager {
-  private readonly apis = observable.map<string, KubeApi>();
-  private readonly stores = observable.map<string, KubeObjectStore>();
+  private readonly externalApis = observable.array<KubeApi>();
+  private readonly externalStores = observable.array<KubeObjectStore>();
 
-  constructor() {
-    makeObservable(this);
-    autoBind(this);
-  }
+  constructor(private readonly dependencies: Dependencies) {}
 
-  getApi(pathOrCallback: string | ((api: KubeApi<KubeObject>) => boolean)) {
-    if (typeof pathOrCallback === "string") {
-      return this.apis.get(pathOrCallback) || this.apis.get(parseKubeApi(pathOrCallback).apiBase);
-    }
+  getApi(pathOrCallback: string | FindApiCallback) {
+    const callback: FindApiCallback = typeof pathOrCallback === "function"
+      ? pathOrCallback
+      : (() => {
+        const { apiBase } = parseKubeApi(pathOrCallback);
+
+        return api => api.apiBase === apiBase;
+      })();
 
-    return iter.find(this.apis.values(), pathOrCallback ?? (() => true));
+    return chain(this.dependencies.apis.get().values())
+      .concat(this.externalApis.values())
+      .find(callback);
   }
 
   getApiByKind(kind: string, apiVersion: string) {
-    return iter.find(this.apis.values(), api => api.kind === kind && api.apiVersionWithGroup === apiVersion);
+    return this.getApi(api => api.kind === kind && api.apiVersionWithGroup === apiVersion);
   }
 
   registerApi<Api>(api: RegisterableApi<Api>): void;
   /**
    * @deprecated Just register the `api` by itself
    */
   registerApi<Api>(apiBase: string, api: RegisterableApi<Api>): void;
-  registerApi<Api>(apiBaseRaw: string | RegisterableApi<Api>, apiRaw?: RegisterableApi<Api>) {
-    const api = typeof apiBaseRaw === "string"
-      ? apiRaw
-      : apiBaseRaw;
-
-    if (!api?.apiBase) {
-      return;
-    }
-
-    if (!this.apis.has(api.apiBase)) {
-      this.stores.forEach((store) => {
-        if (store.api === api) {
-          this.stores.set(api.apiBase, store);
-        }
-      });
-
-      this.apis.set(api.apiBase, api);
+  registerApi<Api>(...args: [RegisterableApi<Api>] | [string, RegisterableApi<Api>]) {
```

Comment:
Nit: Different arguments could be handled cleaner with **method overloading**. Also: why is string-argument present at all if it's not used?

Reply:
backwards compatability and this is method overloading, did you see the diff above.

Reminder: typescript doesn't have true method overloading, but instead it has argument type specifications where the final "concrete" version has to be at least as general as all the "overloads" combined.